### PR TITLE
Work around nvcc_wrapper std::complex problem

### PR DESCRIPTION
**Pull request by masterleinad**
`CUDA` device code is not compatible with `std::complex` and compiling `deal.II` using `nvcc_wrapper` with `DEAL_II_WITH_COMPLEX_VALUES=ON` failed with error messages like
```
/export/home/darndt/dealii/include/deal.II/base/tensor.h(1261): error: a value of type "int" cannot be assigned to an entity of type "_ZNSt7complexIdE9_ComplexTE"
```
This pull request just calls `assert(false)` instead if such an operation is requested.

### CODE REVIEW

**Comment by jppelteret** [comment]
Couldn't you just do something like this:
```c++
multiply_assign_scalar(std::complex<Number> &val, const OtherNumber &s)
  {
    val *= internal::NumberType<Number>::value(s);
  }
```
? I would think that it would be reasonable to scale a complex number by a value of its own type.

**Comment by masterleinad** [comment]
No, you get
```
 /export/home/darndt/dealii/include/deal.II/base/tensor.h(942): error: a value of type "int" cannot be assigned to an entity of type "_ZNSt7complexIdE9_ComplexTE"
```
just as before.

File: include/deal.II/base/tensor.h [hunk 1]
```diff
@@ -924,6 +924,35 @@ Tensor<0, dim, Number>::operator-=(const Tensor<0, dim, OtherNumber> &p)
...
+  {
+    tensor *= s;
+  }
+
+  template <typename Number, typename OtherNumber>
+  DEAL_II_CUDA_HOST_DEV void
+  multiply_assign_scalar(std::complex<Number> &, const OtherNumber &)
+  {
+    printf("This function is not implemented for std::complex<Number>!\n");
+    assert(false);
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: Couldn't you do `Assert (false, ExcMessage(...))` here?

File: include/deal.II/base/tensor.h [hunk 2]
```diff
@@ -924,6 +924,35 @@ Tensor<0, dim, Number>::operator-=(const Tensor<0, dim, OtherNumber> &p)
...
+template <int dim, typename Number>
+template <typename OtherNumber>
+inline DEAL_II_CUDA_HOST_DEV Tensor<0, dim, Number> &
+Tensor<0, dim, Number>::operator*=(const OtherNumber &s)
+{
+  internal::multiply_assign_scalar(value, s);
+  return *this;
+}
+#else
 template <int dim, typename Number>
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: leave some empty lines here to make this easier to read

File: include/deal.II/base/tensor.h [hunk 3]
```diff
@@ -932,6 +961,8 @@ Tensor<0, dim, Number>::operator*=(const OtherNumber &s)
   value *= s;
   return *this;
 }
+#endif
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: Actually, I'd rather just always use the same implementation of the function. Isn't that possible?

File: include/deal.II/base/tensor.h [hunk 4]
```diff
@@ -924,6 +924,35 @@ Tensor<0, dim, Number>::operator-=(const Tensor<0, dim, OtherNumber> &p)
 }
 
 
+
+#ifdef __CUDA_ARCH__
+namespace internal
+{
+  template <typename Number, typename OtherNumber>
+  DEAL_II_CUDA_HOST_DEV void
+  multiply_assign_scalar(Number &tensor, const OtherNumber &s)
```
Comment (bangerth) [inline include/deal.II/base/tensor.h]: mark as inline

**Comment by bangerth** [comment]
I'd also like it better if the function is hidden in a further namespace inside `internal`.

File: include/deal.II/base/tensor.h (hunk 1 above)
Comment (masterleinad) [reply include/deal.II/base/tensor.h]: No, we call `host` functions in `Assert`. We use the exact same workaround in `lac/cuda_kernels.h`.

File: include/deal.II/base/tensor.h (hunk 3 above)
Comment (masterleinad) [reply include/deal.II/base/tensor.h]: Yes, I can just add the `std::complex` overload in case the device path is compiled.

**Pull request merged**